### DISCUSSION TITLE
Why use turbo for the combobox?

### DISCUSSION THREAD

**Discussion by dlegr250** [josefarias/hotwire_combobox#182] [General]
The library looks very promising! I was wondering why use turbo to implement this, though? Wouldn't StimulusJS and returning HTML from the server be enough? You execute an AJAX/fetch/request.js call, get HTML from the server, replace the list innerHTML, and use StimulusJS lifecycle hooks to perform events after the list is updated. Curious what the benefit is for turbo?

**Comment by josefarias** [comment]
Ergonomics! It's the same benefit from using Stimulus over mutation observers and event listeners 😄

In the end, it's ultimately part of the toolbox I enjoy using the most — and what I think most small-team CRUD apps should use. The reasoning behind using Turbo is the same reason for using anything other than vanilla JS for this.

I appreciate the question! But it's not something we'll change so I'll close this for now. But feel free to continue the discussion below.

**Comment by dlegr250** [comment]
I was mostly just curious about the reason for using turbo, didn't want to change how it's used in the gem :)

For context, I had written an autocomplete utility for a project a few years back using just StimulusJS and returning HTML with no layout, and I would just update the list's innerHTML to the updated list items and it all worked really well. I'm currently updating and refactoring those pieces of the project, and wanted to take a look at updating the autocomplete to use turbo and the request.js library.

I ran into issues with the refactoring where the turbo response updates the list results but the way request.js automagically handles turbo responses meant I didn't have clear callbacks I could use to execute actions after the request was finished (request.js apparently doesn't expose the actual promise when handling turbo responses).

So that led me to some googling when I came across this project, and was curious about the architecture decisions being made. It looks really good!

**Reply by josefarias** [reply]
Ah I see! Yeah, what we're doing here is connecting a stimulus target as the last element in the turbo stream response, which signals the end of the stream. Other folks have extended Turbo to include an after-stream-render callback, since I think the promises for rendering streams themselves aren't exposed even at the Turbo level. But I might be misremembering that.

Definitely a challenge!

FWIW I'm seriously considering removing request.js here. It's good, but more than this gem needs, and it adds an undocumented dependency and overhead.
